### DISCUSSION TITLE
How to extract author's family name when adding a reference?

### DISCUSSION THREAD

**Discussion by Klafyvel** [papis/papis#888] [Q&A]
Hi, 
I'm trying papis coming from Zotero. However, I am completely lost on how to have nicely formatted references when I add a document. Say I have the following ref-format config (adapted from the doc):
```
ref-format = "{doc[author][surname]}{doc[year]}{doc[title]:.15}"
```
If I run the following to add a reference, I get an error:
```
$ papis add  paper.pdf --set author "John Doe" --set year 2011 --set title "reference"
[ERROR] paths: Could not format path for document. (Caught exception 'FormatFailedError: {doc[author][surname]}{doc[year]}{doc[title]:.15...'. Use `--log DEBUG` to see traceback)
```
How do I get it to work? I'm okay with writing some smart-ish code to do it for me, but I can't find a list of defined hooks in papis in the [doc](https://papis.readthedocs.io/en/latest/hooks.html) where I could plug some script. Honestly I'm lost here...

**Comment by alexfikl** [comment]
> Hi, I'm trying papis coming from Zotero. 

Thank you for giving it a try! There's a bit of a curve until you get everything configured "just right", but it will hopefully get easier!

> 
> ```
> ref-format = "{doc[author][surname]}{doc[year]}{doc[title]:.15}"
> ```

This seems to be a bit outdated. In the `doc`, the authors are under the `author_list` key and have the format `[{"family": "Doe", "given": "John"}, ...]`. So what you'd want is
```
ref-format = "{doc[author_list][0][family]}{doc[year]}{doc[title]:.15}"
```
However, this can still lead to some errors: the format string is just Python's [str.format](https://docs.python.org/3/library/string.html#formatstrings). For example: if the "author_list" is empty it will raise an `IndexError` or if the first author does not have a "family" it will raise a `KeyError`. This can happen pretty easily depending on where you import your data from and how nicely cleaned up it is.

In general, you can run `papis --log DEBUG add ...` and it will tell you what key it doesn't like in your `ref-format` string.

> How do I get it to work? I'm okay with writing some smart-ish code to do it for me, but I can't find a list of defined hooks in papis in the doc where I could plug some script.

Unfortunately the hook infrastructure is pretty bare bones. We started adding it at some point, but never got it to a point where it's comprehensive and usable.

An "easy" way to hack into the reference building with some code is to define a formatter. E.g. add something like this to your `config.py`:
```
import papis.format

class MyFormatter(papis.format.PythonFormatter):
	def format(self,
               fmt: str,
               doc: papis.document.DocumentLike,
               doc_key: str = "",
               additional: Optional[Dict[str, Any]] = None,
               default: Optional[str] = None) -> str:
		if fmt == "X-REF-FORMAT-X":
			return my_fancy_ref_construction(doc)
		else:
			return super().format(fmt, doc, doc_key=doc_key, additional=additional, default=default)


papis.format.FORMATTER = MyFormatter()
```
and then set `ref-format = X-REF-FORMAT-X` in your configuration file. I think something like this should work, but haven't really tested it :\\

This is all probably a bit much, so feel free to ask as many follow up questions as you need! Papis is pretty flexible, so there's a lot of ways to make it work they way you need it :grin:

**Comment by alexfikl** [comment]
#892 adds a hook that gets called after all the editing is done on `papis add`. Hopefully that's a good time to add a reference too. Any testing would be very welcome!

**Comment by Klafyvel** [comment]
Hi, thank you very much for your answer, and the PR, I'll give it a try asap! 

I'd be happy to contribute to papis or its documentation once I get it to work. I was able to import rather smoothly my database from Zotero by writing [a script](https://github.com/Klafyvel/ZoteroToPapis.jl) but was a bit frustrated when I wrote my question yesterday. Is there a place for informal discussions about papis dev?

**Reply by alexfikl** [reply]
> Is there a place for informal discussions about papis dev?

We mostly use Github discussions for that. We used to have a Zulip chat and some Libera channel, but I think they kind of fizzled out :(

**Reply by alexfikl** [reply]
> I was able to import rather smoothly my database from Zotero by writing [a script](https://github.com/Klafyvel/ZoteroToPapis.jl) [...]

Did you try [papis-zotero](https://github.com/papis/papis-zotero) for that? It's been improving a bit in the last year, but it's hard to keep it up to date since it's usually a "once and done" sort of usage.

**Reply by Klafyvel** [reply]
I tried briefly, but I wasn't super happy with how it named and placed the references, and I think it did not handle better-bibtex very well (although it may be a mistake on my end).

**Reply by alexfikl** [reply]
> I tried briefly, but I wasn't super happy with how it named and placed the references

Hm, it should just get the metadata from Zotero and then pass it on to `papis add` which constructs the refs in the same way (using `ref-format` and friends). If you ever get back to some of that and have some time to file some issues, that would be very welcome!

> I think it did not handle better-bibtex very well (although it may be a mistake on my end).

What special handing does better-bibtex need?

**Reply by Klafyvel** [reply]
>  If you ever get back to some of that and have some time to file some issues, that would be very welcome!

I'll have a look.

> What special handing does better-bibtex need?

The citation keys are stored in another database, it's fairly simple to fetch though:https://github.com/Klafyvel/ZoteroToPapis.jl/blob/main/src/database_queries.jl#L135
